src/Stratis.Bitcoin/Utilities/AsyncQueue.cs

```diff
@@ -93,11 +115,101 @@ public void Enqueue(T item)
             }
         }
 
+        /// <summary>
+        /// Dequeues an item from the queue if there is one. 
+        /// If the queue is empty, the method waits until an item is available.
+        /// </summary>
+        /// <param name="cancellation">Cancellation token that allows aborting the wait if the queue is empty.</param>
+        /// <returns>Dequeued item from the queue.</returns>
+        /// <exception cref="OperationCanceledException">Thrown when the cancellation token is triggered or when the queue is disposed.</exception>
+        /// <exception cref="InvalidOperationException">Thrown if this method is called on a queue that operates in callback mode.</exception>
+        public async Task<T> DequeueAsync(CancellationToken cancellation = default(CancellationToken))
+        {
+            if (this.callbackMode)
+                throw new InvalidOperationException();
```

Comment:
I think it will be a good idea to let user know why we're throwing this. Maybe add a message like: 
"Dequeue is not supported if 'onEnqueue' callback is specified."

Reply:
ok, will add message in this exception, tx